### PR TITLE
ref: add haskell implementation

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

File: ref/haskell/src/Lib.hs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+bech32Encode :: HRP -> Data -> Maybe BS.ByteString
+bech32Encode hrp dat = do
+    let dat' = dat ++ bech32CreateChecksum hrp dat
+    rest <- mapM (flip Map.lookup charsetMapInv) dat'
+    return $ BSC.concat [hrp, "1", BSC.pack rest]
+
+bech32Decode :: BS.ByteString -> Maybe (HRP, Data)
+bech32Decode bech32 = do
+    guard $ BS.length bech32 <= 90
+    guard $ BSC.map toUpper bech32 == bech32 || BSC.map toLower bech32 == bech32
```
Comment (sipa) [inline ref/haskell/src/Lib.hs]: It seems that toLower/toUpper will map some non-ASCII characters to ASCII characters (in particular 304, 305, 383, and 8490), so I worry that this may cause those to be unintentionally accepted.

Comment (benma) [inline ref/haskell/src/Lib.hs]: [BSC.map](https://hackage.haskell.org/package/bytestring-0.10.8.1/docs/Data-ByteString-Char8.html) works on 8-bit chars, so those are not affected:

> More specifically these byte strings are taken to be in the subset of Unicode covered by code points 0-255. This covers Unicode Basic Latin, Latin-1 Supplement and C0+C1 Controls.

Also, in the same function below there are more validity checks (hrp chars are between 33 and 126, and the data chars are in the base32 charset).

File: ref/haskell/src/Lib.hs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: I could, but Data.IntMap and Data.Map are in a different dependency `containers`. Same for vectors and arrays. I could use lists (standard Haskell lists are linked lists :cry:), but access would be O(n) then. Both `containers` and `unordered-containers` (which I use here) are part of Haskell Platform though, which means it is commonly used (or at least deemed to be).

Comment (sipa) [inline ref/haskell/src/Lib.hs]: Ok.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+To build and run the tests, you need [stack](https://docs.haskellstack.org/en/stable/README/).
```
Comment (sipa) [inline README.md]: Perhaps this file was intended to go in the ref/haskell directory rather than the root?

Comment (benma) [inline README.md]: Indeed. Fixed!

**Comment by benma** [comment]
@roconnor-blockstream thanks for the detailed review! I added individual commits for feedback for easier review, I will squash them in the end.

> I think the safety of this module could be greatly improved by creating a `newtype Word5 = UnsafeWord5 Word8`

Good idea, added in the last commit. I added it to the type signature of bech32(En|De)code, it all comes together nicely. I had to expose the type though to make it available in the tests. If you prefer, I can move all of that to an `Internal` package and only expose the segwit functions in the main module.

File: ref/haskell/src/Lib.hs [hunk 4]
```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: I did that, and the overloaded strings were especially nice in the Spec.hs (unit tests). I am not sure though if it is worth it. As I understand it, OverloadedStrings advanced fast to one of the most used extensions. Not sure which other compilers there that do not implement this extension, but the others that ByteString etc. rely on. (BangPatterns, for example).

File: ref/haskell/src/Lib.hs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+module Lib
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: Moved to Codec.Binary.Bech32.

File: ref/haskell/src/Lib.hs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Map.HashMap Char Word8
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: I could do it for charsetMapInv, but for charsetMap, it would result in a mess. An array needs bounds and be filled continuously within them, but the base32 charset (sorted) is not continuous. This would make the code less elegant, also in base32Decode where I'd need to add explicit charset membership checks.

I could make a newtype and define Ix myself to make it continuous, but that would add a lot of code as well.

Thoughts?

File: ref/haskell/src/Lib.hs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+charsetMap = Map.fromList $ zip charset [0..]
+
+charsetMapInv :: Map.HashMap Word8 Char
+charsetMapInv = Map.fromList $ zip [0..] charset
+
+bech32Polymod :: Data -> Int
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: Done

File: ref/haskell/src/Lib.hs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+bech32Polymod = foldl' go 1
+  where
+    go chk value = foldl' xor chk' [(- (top .>>. i) .&. 1) .&. g | (g, i) <- zip generator [0..]]
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: Nice one! Done.

File: ref/haskell/src/Lib.hs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      where
+        generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3]
+        top = chk .>>. 25
+        chk' = (chk .&. 0x1ffffff) .<<. 5 `xor` (fromIntegral value)
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: As far as I can tell, `Word32` doesn't matter (if the type is bigger, it would just fill the higher bits and discard them later).

Even so, I tried your suggestion, and weirdly, the unit tests fail:

```
Addresses:         FAIL
    "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4"
    expected: Just "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"
     but got: Just "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kq8f3t4"
```

Same error happens in the python ref unit tests when I apply the same change. I don't quite see why yet, seems like it should work.

File: ref/haskell/src/Lib.hs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+bech32HRPExpand :: HRP -> Data
+bech32HRPExpand hrp = map (.>>. 5) (BS.unpack hrp) ++ [0] ++ map (.&. 31) (BS.unpack hrp)
+
+bech32CreateChecksum :: HRP -> Data -> Data
+bech32CreateChecksum hrp dat = [fromIntegral $ polymod .>>. (5 * (5 - i)) .&. 31 | i <- [0..5]]
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: Done. The same could also be done in the python ref implementation.

File: ref/haskell/src/Lib.hs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  where
+    values = bech32HRPExpand hrp ++ dat
+    polymod = bech32Polymod (values ++ [0, 0, 0, 0, 0, 0]) `xor` 1
+
+bech32VerifyChecksum :: HRP -> Data -> Bool
+bech32VerifyChecksum hrp dat = bech32Polymod (bech32HRPExpand hrp ++ dat) == 1
+
+bech32Encode :: HRP -> Data -> Maybe BS.ByteString
+bech32Encode hrp dat = do
```
Comment (benma) [inline ref/haskell/src/Lib.hs]: Done, and added a test for it.

File: ref/haskell/src/Lib.hs (hunk 9 above)
Comment (roconnor-blockstream) [inline ref/haskell/src/Lib.hs]: Sorry I was a bit hasty.

The transformation is 
```
(chk .&. 0x1ffffff) .<<. 5
```
 = { distribution of .<<. and .&. }
```
(chk .<<. 5) .&. (0x1ffffff .<<. 5)
```
 = { compute (0x1ffffff .<<. 5) }
```
(chk .<<. 5) .&. (0x3fffffe0)
```
= { low bits are 0 anyways }
```
(chk .<<. 5) .&. (0x3fffffff)
```

And now you can float the `0x3fffffff` to the top.

```
bech32Polymod = foldl' go 1 .&. 0x3fffffff
```

I think you are right that the `Word32`ness doesn't matter.  I just didn't want to worry about sign extensions when thinking about it.

Comment (benma) [inline ref/haskell/src/Lib.hs]: Of course! Thanks. Pushed in a new commit.

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+instance Hashable Word5 where
+    hashWithSalt salt (UnsafeWord5 x) = hashWithSalt salt x
+
+word5 :: Word8 -> Word5
+word5 x = UnsafeWord5 (x .&. 31)
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+
```
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: `enumFromThen (UnsafeWord5 n1) (UnsafeWord5 n2) = map UnsafeWord5 [n1,n2 .. if (n1 <= n2) then 31 else 0]`

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
```
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]:
```
charset :: Array Word5 Char
charset = listArray (0,31) "qpzry9x8gf2tvdw0s3jn54khce6mua7l"

charsetMap :: Char -> Maybe Word5
charsetMap c | upperC `inRange` bounds invArray = invArray ! upperC
             | otherwise = Nothing
 where
  upperC = toUpper c
  invArr :: Array Char (Maybe Word5)
  invArr = accumArray ('0','Z') (const Just) Nothing (map swap (assocs charset))
   where
    swap (a,b) = (b,a)

charsetMapInv :: Word5 -> Char
charsetMapInv i = charset ! i

Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Oops
`swap (i,c) = (toUpper c, i)`

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 12 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: I do like the sentiment of making the Word5 type flexible and more usable. However, I don't think it is worth it to add this many lines and all the instances just to be able to remove a couple of `UnsafeWord5` calls. After all, this is a bech32 library, not a ShortWord library :)

Also, what do you think of using one of the existing shortword libraries, like https://hackage.haskell.org/package/OddWord? (I just found it, haven't checked in detail if it fits the bill).

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 14]
```diff
@@ -0,0 +1,128 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  , Word5(UnsafeWord5)
```
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Done

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 13 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: I have pushed a commit that does this just to see. Could you explain why this is better? Your `charsetMap` is the equivalent of a the Data.HashMap. Access is constant time in either case (the Hashable instance of Char is just `fromEnum`, so no collisions there either). Memory usage is similar. But now, there is this extra non-trivial code. The HashMap version was very easy to read.

Would you be much opposed to just leaving the HashMap version as it was? This might be premature optimization. There is much lower hanging fruit in the base32->base256 conversion.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 12 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Added the word5 fromIntegral with the specialization.

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: So using Word5 from the OddWord library worked like a charm. See https://github.com/sipa/bech32/pull/11/commits/f730a365f6c716c14fd47a89ebad0b4cc99d1b0a

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 13 above)
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: The hash-map uses `IntMap` as it's implementation which is a Patricia trie. To look up an item in a Patrica trie will requires traversing several indirections.  On top of that there is going to be quite a bit of overhead in the Patrica trie structure itself.  Even with the 11 or so empty cells in the array representation, the memory use is probably going to be less than the Patrica trie.  The array lookup is just an offset and and at most a single indirection after that.  Plus we get the very minor bonus of dropping one popular, but large dependency since Arrays are part of the language standard.

If you are worried about complexity, you can add an exhaustive test case in your set of tests:

```
charsetMap_test :: Bool
charsetMap_test = all correct1 [char i | i <- [0..255]] && all correct2 [word5 i | i <- [0..31]]
  where
    correct1 c = fromMaybe c (charsetMapInv <$> charsetMap c) = c
    correct2 w = charsetMap (charsetMapInv w) = Just w
```

That said, if you really really want `HashMap` I'll concede.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 12 above)
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: I looked at the `OddWord` library, and I'm not very keen on using it.  I immediately noticed that the `succ` and `pred` definitions are incorrect, and you've already noticed that it is missing `Ix` instances.

I can agree that all the instances I wrote out is maybe too much for removing a few uses of UnsafeWord5.

I suggest that you avoid `OddWord` and write

```
newType Word5 = UnsafeWord5 Word8 deriving (Eq, Ord)
```

Export the `word5` and `fromWord5` functions.  Add the `Ix` instance, and, at your discretion, maybe the `Num` instance, and leave it at that.  I suggest you stick the instances at the end of the file, because the instances are obvious, and no one will care to look at them.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 13 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Oh, I opened the code, but drew my conclusions to fast ;) Thanks for the info.

> If you are worried about complexity, you can add an exhaustive test case in your set of tests:

It is not the complexity (I am confident that the code is correct, and it's also indirectly tested by the current tests). Just that someone reading the code will not immediately understand it, unlike the HashMap version.

I left the array based implementation now, though, as I agree that arrays are technically superior.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 12 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Ok, then it's settled ;) I removed the OddWord commit again, and didn't add any instances apart from Ix.

**Comment by benma** [comment]
@roconnor-blockstream thanks again, this was pretty awesome! 

I squashed everything into one commit.

After this is merged, the next step is to replace the base conversion functions with something more efficient, not using the digits library and Integer.

File: ref/haskell/stack.yaml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+resolver: lts-8.5
+packages:
+- '.'
+extra-deps: [OddWord-1.0.1.0]
```
Comment (roconnor-blockstream) [inline ref/haskell/stack.yaml]: OddWord can be removed now I think.

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+charsetMap :: Char -> Maybe Word5
+charsetMap c | inRange (Arr.bounds inv) upperC = inv Arr.! upperC
+             | otherwise = Nothing
+  where
+    upperC = toUpper c
+    inv = Arr.listArray ('0', 'Z') (repeat Nothing) Arr.// (map swap (Arr.assocs charset))
+    swap (a, b) = (toUpper b, Just a)
+
+bech32Polymod :: [Word5] -> Word32
```
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: One last comment.  I know I said to have `bech32Polymod` return a `Word32`, however now that I've studied the issue, if you think it is better, you could return `Word`.  My understanding of the Haskell report is that `Data.Word.Word` is guaranteed to be at least 30 bits (although this isn't stated quite as clearly as I would like).  Using `Int` as in your original code does work, but I feel that all bit-shifting work should be done on unsigned words to avoid the cognitive load of thinking about how sign-extension may or may not interact with the code.

Long story short, you can return `Word` if you think it is better.  I think both `Word` or `Word32` are fine.

File: ref/haskell/stack.yaml (hunk 15 above)
Comment (benma) [inline ref/haskell/stack.yaml]: Good catch, thanks. Done.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 16 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: So what would be the advantage of using Word over Word32? Isn't it also a cognitive load to wonder how many bits Word has and whether it's enough, whereas Word32 is clear?

How many bits are actually needed, 29?

**Comment by sipa** [comment]
30 bits are needed (6 checksum characters of 5 bits each)

**Comment by benma** [comment]
Of course, `0x3fffffff` appears even in the function :see_no_evil:

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 16 above)
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: The possible advantage of using `Word` over `Word32` is that `Word` is (likely) a native machine word and should have slightly faster operations.  `Word` is also slightly less exotic than `Word32`.

**Comment by roconnor-blockstream** [comment]
sipa was smart enough to have a checksum size that fits in ML-sized integers that reserve 2-bits for tagging. :D

**Comment by sipa** [comment]
> a checksum size that fits in ML-sized integers that reserve 2-bits for tagging

yes, clearly a design goal.

File: ref/haskell/src/Codec/Binary/Bech32.hs (hunk 16 above)
Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Done, changed to `Word`.

**Comment by benma** [comment]
@roconnor-blockstream, I also made much faster base conversion functions by getting rid of the `digits` dependency. It works pretty much like the `convertbits` of the Python implementation. Please take a look.

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 17]
```diff
@@ -97,30 +97,32 @@ bech32Decode bech32 = do
     guard $ bech32VerifyChecksum hrp' dat'
     return (hrp', take (BS.length dat - 6) dat')
 
-toBase32 :: [Word8] -> [Word5]
-toBase32 dat = pad ++ result
+convertBits :: [Word] -> Int -> Int -> Bool -> Maybe [Word]
```
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: I'd like a little documentation about covertBits as a comment:
* frombits and twobits must be positive and less than the bitSize of Word.
* Every value of Word must be strictly less than 2^frombits.
* The conversion is Big Endien.

Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: There is a way to remove the use of `fromJust` but you might not like it.

```
type Pad f = Int -> [Word] -> [[Word]] -> f [[Word]]

yesPadding :: Pad Identity
yesPadding _ bits padValue result | bits /= 0 = return $ [padValue] : result
                                  | otherwise = return result
{-# INLINE yesPadding #-}

noPadding :: Pad Maybe
noPadding fromBits bits padValue result = do
  guard $ bits < fromBits && padValue == 0
  return result
{-# INLINE noPadding #-}

convertBits :: Functor f => [Word] -> Int -> Int -> Pad f -> f [Word]
{- ... -}
  go [] acc bits result = pad fromBits bits ((acc .<<. (tobits - bits)) .&. maxv) result
{- ... -}

toBase32 :: [Word8] -> [Word5]
toBase32 dat = map word5 $ runIdentity $ convertBits (map fromIntegral dat) 8 5 yesPadding
  
toBase256 :: [Word5] -> Maybe [Word8]
toBase256 dat = fmap (map fromIntegral) $ convertBits (map fromWord5 dat) 5 8 noPadding
```

If this technique excites you can use it, but I won't force it on you.  What you have already written is acceptable.

[#BooleanBlindness](https://shreevatsa.wordpress.com/2015/01/31/boolean-blindness/)

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: It indeed felt unsettling to use the partial fromJust, and using Identity to get rid of it in fact crossed my mind, but I abstained because I felt it might be overkill (fromJust being safe here was easy to see/prove after all).

I think I'll adopt your suggestion anyway, it is hard to resist type safety :) 

About boolean blindness: I also considered just doing `data Padding = Pad | NoPad`. Here I actually don't know why I didn't do it in the end :see_no_evil:

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Done.

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Pushed.

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 18]
```diff
@@ -11,14 +11,14 @@ module Codec.Binary.Bech32
   ) where
 
 import Control.Monad (guard)
+import Control.Monad.Identity (Identity, runIdentity)
```
Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Data.Functor.Identity is in base I believe, and you won't need the mtl.

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Cool! Done.

File: ref/haskell/src/Codec/Binary/Bech32.hs [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+  where
+    upperC = toUpper c
+    inv = Arr.listArray ('0', 'Z') (repeat Nothing) Arr.// (map swap (Arr.assocs charset))
+    swap (a, b) = (toUpper b, Just a)
+
+bech32Polymod :: [Word5] -> Word
+bech32Polymod values = foldl' go 1 values .&. 0x3fffffff
+  where
+    go chk value = foldl' xor chk' [g | (g, i) <- zip generator [25..], testBit chk i]
+      where
```
Comment (sipa) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Is there a reason to have a `where` inside a `where`?

Comment (roconnor-blockstream) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: `chk'` depends on `go`'s `chk` parameter, so that has to be nested.  The `generator` could be lifted out one level, but it is better to keep the scope of things as small as possible.

Comment (benma) [inline ref/haskell/src/Codec/Binary/Bech32.hs]: Just clearer scoping, i.e. you can't use `generator` and `chk'` in line 61.

**Comment by benma** [comment]
@roconnor-blockstream @sipa squashed.

**Comment by roconnor-blockstream** [comment]
This looks good to me.  I approve.

**Comment by benma** [comment]
@roconnor-blockstream thanks again for the thorough review, I enjoyed it a lot and learned a thing or two!
